source/_static/lecture_specific/opt_tax_recur/crra_utility.py

```diff
@@ -1,38 +1,43 @@
-import numpy as np
-
+crra_util_data = [
+    ('β', float64),
+    ('σ', float64),
+    ('γ', float64)
+]
 
+@jitclass(crra_util_data)
 class CRRAutility:
 
     def __init__(self,
                  β=0.9,
                  σ=2,
-                 γ=2,
-                 π=0.5*np.ones((2, 2)),
-                 G=np.array([0.1, 0.2]),
-                 Θ=np.ones(2),
-                 transfers=False):
+                 γ=2):
 
         self.β, self.σ, self.γ = β, σ, γ
-        self.π, self.G, self.Θ, self.transfers = π, G, Θ, transfers
 
     # Utility function
-    def U(self, c, n):
+    def U(self, c, l):
         σ = self.σ
         if σ == 1.:
             U = np.log(c)
         else:
             U = (c**(1 - σ) - 1) / (1 - σ)
-        return U - n**(1 + self.γ) / (1 + self.γ)
+        return U - (1-l) ** (1 + self.γ) / (1 + self.γ)
```

Comment:
My understanding is that `n+l=1` is not supposed to be enforced for this case (at least with with `n>=0` and `l>=0`). Quoting from Tom's email: "It is ok if [n] exceeds 1; this part of the lecture is using the alternative utility and technology specification". Of course, you can define `l` this way, but then, I think it won't match what labor is supposed to be under this "alternative utility and technology specification". 

More generally, at least in the AMSS lecture, there's a mismatch between the model being solved and the model described in the lecture. Is this the case here too?

Reply:
I am not sure about the `amss` lecture but here `n+l=1` is enforced (see equation (1)).

Reply:
What about the "Specification with CRRA Utility" section that says "We also eliminate Lucas and Stokey’s restriction that ℓt+nt≤1."

Reply:
Note that right below the sentence you quoted, it says "With these adjustments, the analysis of Lucas and Stokey prevails once we make the following replacements".

To me the original complete formulation of the model from eq. (1)-(26) is what I should follow closely when we design the code, and the CRRA case which eliminates "l+n=1" is just a special case and special replacements are needed to fit it into the original general framework.

Since I am not imposing a restriction on l, the possible range for n in the CRRA case after replacements should still be [0,+∞], and imposing `n+l=1` is no more than a linkage of a special specification to the general framework defined above.

Reply:
Yes, I understand. My point is just that you'll get `l<0` since `n>1`. I don't think the lecture specifies `l<0`, but negative leisure is surprising. In the LS paper, the utility function seems to be a mapping from R^2_{+} to R (below eq 2.2), so it looks like they don't allow for negative leisure although I would probably need to read the paper carefully to verify this. Hence, my guess is that the `n` in the CRRA specification is not the `n` in (1).

Reply:
Since the CRRA formulation eliminates `l`, it is probably not so meaningful to discuss the sign of `l`.

I may want to clarify my two reasons for designing the code like this: (1) I prefer to impose `n+l=1` in the class, since this is how it is defined at the very beginning when introducing the general framework, (2) we want to fit the CRRA case into the general framework, in that case `n=1-l` is necessary. Since the "economic" meaning of `l` has been eliminated in the CRRA case, `l` plays a role no more than linking the specific CRRA case to the general framework, and a negative `l` associated with `n>1` does not bother me too much.

Might this sound a little more reasonable to you?

Reply:
I understand that you want to fit the CRRA case into the "general" framework. I am just wondering whether imposing `n=1-l` is the right way to do it. I'll double-check with Tom.

Reply:
Sure it would be great to consult Tom about this. Please include me in the communication.